loader.php

```diff
@@ -15,45 +15,129 @@ function is_development() {
 }
 
 /**
- * Get the port for React's development server.
+ * Create a probably-unique key for a given script bundle.
  *
- * @return int|null Port number if available, otherwise null.
+ * @param string $path A relative filesystem path or resource URI.
+ * @return string
  */
-function get_react_port() {
-	if ( ! is_development() ) {
-		return null;
+function get_asset_key( string $path ) {
+	$path_parts = explode( '/', $path );
+	return 'react_wp_script_' . end( $path_parts );
+}
+
+/**
+ * Convert a relative path within a theme or plugin to an absolute filesystem path.
+ *
+ * @param string $path     A relative file system path.
+ * @param bool   $is_theme Whether to treat the $path as a theme file.
+ * @return string;
+ */
+function get_file_path( string $path, bool $is_theme ) {
+	return $is_theme
+		? get_theme_file_path( $path )
+		: plugin_dir_path( __file__ ) . $path;
+}
+
+/**
+ * Check the filesystem for asset manifests for a theme or plugin and attempt to
+ * decode and return the asset list JSON if found.
+ *
+ * @param bool $is_theme Whether to look for the manifest in a theme or plugin context.
+ * @return array|null;
+ */
+function get_assets_list( string $is_theme ) {
```

Comment:
This function feels clunky but it's the only way I could see to achieve the desired fallback order of

1. If development, try to load and enqueue assets from dev server bundle
2. If not development or dev server assets bundle fails to load, try to load and enqueue assets from built bundle